packages/neo-one-monitor/src/BrowserMetricsFactory.js

```diff
@@ -0,0 +1,144 @@
+    this.metric = metric;
+    this.values = [];
+  }
+
+  reset(): void {
+    this.values = [];
+  }
+}
+
+export type MetricCollection = {|
+  counters: { [name: string]: CollectingMetric },
+  histograms: { [name: string]: CollectingMetric },
+|};
+
+class BrowserCounter extends CollectingMetric implements Counter {
+  inc(countOrLabels?: number | MetricLabels, count?: number): void {
+    this.values.push({ countOrLabels, count });
+  }
+}
+
+export class BrowserGauge extends CollectingMetric implements Gauge {
+  // eslint-disable-next-line
+  inc(countOrLabels?: number | MetricLabels, count?: number): void {}
+
+  // eslint-disable-next-line
+  dec(countOrLabels?: number | MetricLabels, count?: number): void {}
+
+  // eslint-disable-next-line
+  set(countOrLabels?: number | MetricLabels, count?: number): void {}
+}
+
+class BrowserHistogram extends CollectingMetric implements Histogram {
+  observe(countOrLabels?: number | MetricLabels, count?: number): void {
+    this.values.push({ countOrLabels, count });
+  }
+}
+
+export class BrowserSummary extends CollectingMetric implements Summary {
+  // eslint-disable-next-line
+  observe(countOrLabels?: number | MetricLabels, count?: number): void {}
+}
+
+export default class BrowserMetricsFactory implements MetricsFactory {
+  collect(): MetricCollection {
+    const currentMetrics = this._serializeJSON(metrics);
+    for (const metricType of Object.keys(metrics)) {
+      for (const name of Object.keys(metrics[metricType])) {
+        metrics[metricType][name].reset();
+      }
+    }
+    return currentMetrics;
+  }
+
+  createGauge(options: MetricConstruct): Gauge {
+    return new BrowserGauge(options);
+  }
+
+  createCounter(options: MetricConstruct): Counter {
+    const counter = new BrowserCounter(options);
+    metrics.counters[counter.metric.name] = counter;
+
+    return counter;
+  }
+
+  createHistogram(options: MetricConstruct): Histogram {
+    const histogram = new BrowserHistogram(options);
+    metrics.histograms[histogram.metric.name] = histogram;
+
+    return histogram;
+  }
+
+  createSummary(options: MetricConstruct): Summary {
+    return new BrowserSummary(options);
+  }
+
+  _serializeJSON(currentMetrics: MetricCollection): MetricCollection {
+    const copy = {
+      counters: {},
+      histograms: {},
+    };
+
+    for (const metricType of Object.keys(currentMetrics)) {
+      for (const name of Object.keys(currentMetrics[metricType])) {
+        const metricCopy = {
+          metric: {},
+          values: [],
+        };
+        metricCopy.metric.name = currentMetrics[metricType][name].metric.name;
+        metricCopy.metric.help = currentMetrics[metricType][name].metric.help;
+        metricCopy.metric.labelNames = [];
+        currentMetrics[metricType][name].metric.labelNames.map(label =>
+          metricCopy.metric.labelNames.push(label),
+        );
+        metricCopy.values = [];
+        currentMetrics[metricType][name].values.forEach(value => {
```

Comment:
This is a perfect example of what I mean by data transformations vs mutations. As a reader, I now have to look through all of this code line by line to figure out what the overall goal is here of looping over the values. If instead, you told me up front what the goal is by using a transformation function, i.e. `map`, I immediately know we're transforming the current metrics' values. Instead, I have to read through this and more manual iteration to figure that out.